### ISSUE TITLE
Bug router angular v5.2.4+ with cordova (BrowserAnimationsModule bug)

### ISSUE THREAD

**Issue by wilker7ribeiro** [angular/angular#22509]
This bug was introduced in version 5.2.4.
The hardware back button (navigate back) breaks the page, sometimes crashing the application when changing routes.
This bug doesn't happens in v5.2.3 and before.

Obs: only happens if including cordova.js inside index.html

Steps:
**Install cordova
clone project: https://github.com/wilker7ribeiro/bug-cordova-angular-route.git**
**Connect android device**
npm install
ng build
cordova platform add android
cordova run android
**Navigate a little
Press phone hardware backbutton**

**Comment by jasonaden** [comment]
This is an interesting bug. Looking through the differences between `5.2.3` and `5.2.4` ([here](https://github.com/angular/angular/compare/5.2.3...5.2.4)), I can't see what might cause this problem.

Also, it sounds like you have a workaround for the issue. Is that correct? It would be good to get more detail on exactly how to reproduce this issue as well as info on how to make it work under `5.2.4` (sounds from the above like including `cordova.js` in `index.html` causes the issue -- what's the alternative and how do we set that up?).

**Comment by wilker7ribeiro** [comment]
Under 5.2.4 it works in any situation (with or without `cordova.js`).
Sounds like 5.2.4 introduced the issue.

`cordova.js`  give access to cordova and cordova-plugins api (bridge from native android to javascript)

The issue happens with 5.2.4 +  `cordova.js`.
Sounds like it is some incompatibility

I don't have any workaround.
If I remove `cordova.js` from `index.html` it will work (webapp deployed as mobile app, and navigation without issues), but I wont be able to use cordova plugins for native device access (like geolocation, camera, etc).

The project already replicates the issue
To remove the issue just downgrade angular packages to 5.2.3.

Issue behavior:
Router-outlet is appending instead of replacing components, messing up the router and data bindings, some times breaking the app

Maybe it has some relation with [this issue](https://github.com/angular/angular/issues/20290) and [this change](https://github.com/angular/angular/compare/5.2.3...5.2.4#diff-91a41028732ff4a368a53bc46649eb79L312)

Tonight I will see if removing BrowserAnimationsModule solves the issue

**Comment by wilker7ribeiro** [comment]
To reproduce:
Cordova and android environment needs to be correctly [installed]
I'm using **cordova 8.0.0** and **cordova-android 7.0.0**
(https://cordova.apache.org/docs/en/latest/guide/platforms/android/):
Clone: https://github.com/wilker7ribeiro/bug-cordova-angular-route.git
npm install
Execute `cordova platform add android` to configure android platform
Execute `ng build` to build source code into www
Execute `cordova run android` to deploy the app into your android device
Basically there are 3 links to navigate:
/
--/home
--/login
Just go back and forth to create some navigation history
Then press back button on your cellphone to navigate back (**this trigger the issue**)
Some things will happen:
Components will be stacking up when navigating back and forth, instead of replacing each other
If the component has bindings it will mess up
Eventually the app will crash

**Comment by wilker7ribeiro** [comment]
And the answer is Yes!
**BrowserAnimationsModule**  is causing the issue.
If I remove it from AppModule everything works fine

**Comment by gabrie-allaigre** [comment]
Hi,
I same problems.
Work with 5.2.3 but not working with 5.2.4+

`location.back()` have same problems.

Thanks

**Comment by VladyslavKurmaz** [comment]
Hi.
I have the same issue with v5.2.0 and cordova, but without BrowserAnimationsModule - **push back "hardware" button**
Plus, the issue is reproducible when I try to use FB login JS SDK - **access typescript object inside FB API callback**
Looks like, this is similar issue:
https://github.com/angular/angular/issues/20112

---

bash-3.2$ ng --version
```
Angular CLI: 1.7.3
Node: 8.9.1
OS: darwin x64
Angular: 5.2.7
... animations, common, compiler, compiler-cli, core, forms
... http, language-service, platform-browser
... platform-browser-dynamic, router

@angular/cli: 1.7.3
@angular-devkit/build-optimizer: 0.3.2
@angular-devkit/core: 0.3.2
@angular-devkit/schematics: 0.3.2
@ngtools/json-schema: 1.2.0
@ngtools/webpack: 1.10.2
@schematics/angular: 0.3.2
@schematics/package-update: 0.3.2
typescript: 2.5.3
webpack: 3.11.0
```
bash-3.2$ cordova -v
```
8.0.0
```
bash-3.2$ cat package.json
```
{
  "name": "attlas-static-client",
  "version": "18.3.0-SNAPSHOT",
  "license": "MIT",
  "scripts": {
    "ng": "ng",
    "start": "ng serve",
    "build": "ng build --prod",
    "test": "ng test",
    "lint": "ng lint",
    "e2e": "ng e2e"
  },
  "private": true,
  "dependencies": {
    "@angular/animations": "^5.2.0",
    "@angular/common": "^5.2.0",
    "@angular/compiler": "^5.2.0",
    "@angular/core": "^5.2.0",
    "@angular/forms": "^5.2.0",
    "@angular/http": "^5.2.0",
    "@angular/platform-browser": "^5.2.0",
    "@angular/platform-browser-dynamic": "^5.2.0",
    "@angular/router": "^5.2.0",
    "@ng-bootstrap/ng-bootstrap": "^1.0.0",
    "@types/facebook-js-sdk": "^2.8.3",
    "bootstrap": "^4.0.0",
    "core-js": "^2.4.1",
    "intl": "^1.2.5",
    "rxjs": "^5.5.6",
    "zone.js": "^0.8.19"
  },
  "devDependencies": {
    "@angular/cli": "^1.7.3",
    "@angular/compiler-cli": "^5.2.0",
    "@angular/language-service": "^5.2.0",
    "@types/jasmine": "~2.8.3",
    "@types/jasminewd2": "~2.0.2",
    "@types/node": "~6.0.60",
    "codelyzer": "^4.0.1",
    "jasmine-core": "~2.8.0",
    "jasmine-spec-reporter": "~4.2.1",
    "karma": "~2.0.0",
    "karma-chrome-launcher": "~2.2.0",
    "karma-coverage-istanbul-reporter": "^1.2.1",
    "karma-jasmine": "~1.1.0",
    "karma-jasmine-html-reporter": "^0.2.2",
    "karma-phantomjs-launcher": "^1.0.4",
    "protractor": "~5.1.2",
    "ts-node": "~4.1.0",
    "tslint": "~5.9.1",
    "typescript": "~2.5.3"
  }
}
```

**Comment by Billskate** [comment]
Hi, guys
I found the solution by doing few monkey-patching **before** import `cordova.js`.
See also: [this blog](http://weblogs.thinktecture.com/thomas/2017/02/cordova-vs-zonejs-or-why-is-angulars-document-event-listener-not-in-a-zone.html).
All in all, the monkey-patching code looks like:
```html
<script>
  window.addEventListener = function () {
    EventTarget.prototype.addEventListener.apply(this, arguments);
  };
  window.removeEventListener = function () {
    EventTarget.prototype.removeEventListener.apply(this, arguments);
  };
  document.addEventListener = function () {
    EventTarget.prototype.addEventListener.apply(this, arguments);
  };
  document.removeEventListener = function () {
    EventTarget.prototype.removeEventListener.apply(this, arguments);
  };
</script>
<script src="cordova.js"></script>
```

**Comment by VladyslavKurmaz** [comment]
@Billskate works for me, thanks!

**Comment by sat-m** [comment]
The solution by @Billskate  does not work in Internet Explorer (EventTarget is undefined)  :( 
Had do downgrade to 5.2.3 to fix the issue

**Comment by Billskate** [comment]
@sat-m Sorry, I didn't test in IE.
If there is not ```EventTarget```, zone.js will patch ```Window.prototype.addEventListener```.
So the code may looks like:
```html
<script>
  window.addEventListener = function () {
    (window.EventTarget || Window).prototype.addEventListener.apply(this, arguments);
  };
  window.removeEventListener = function () {
    (window.EventTarget || Window).prototype.removeEventListener.apply(this, arguments);
  };
  document.addEventListener = function () {
    (window.EventTarget || Document).prototype.addEventListener.apply(this, arguments);
  };
  document.removeEventListener = function () {
    (window.EventTarget || Document).prototype.removeEventListener.apply(this, arguments);
  };
</script>
<script src="cordova.js"></script>
```
It may also works on IOS safari before 10.1 that have no ```EventTarget```.

**Comment by sat-m** [comment]
Thank you @Billskate, will try this out and get back with the results :)

**Comment by Thomas-Webber** [comment]
@sat-m @Billskate  We should import this only for cordova

```javascript
<!-- Include Cordova script only when needed -->
<script>
  if (!/^http/.test(location.protocol)|/^localhost:8080/.test(location.host)) {
    window.addEventListener = function () {
        EventTarget.prototype.addEventListener.apply(this, arguments);
    };
    window.removeEventListener = function () {
        EventTarget.prototype.removeEventListener.apply(this, arguments);
    };
    document.addEventListener = function () {
        EventTarget.prototype.addEventListener.apply(this, arguments);
     };
    document.removeEventListener = function () {
        EventTarget.prototype.removeEventListener.apply(this, arguments);
    };
    document.write('<script src="cordova.js"><\/script>');
  }
</script>
```

**Comment by h-arora** [comment]
@Billskate This works like magic. I am not an expert in angular so not sure if it has any other impact/side-effect?

**Comment by richenyadav001** [comment]
After update my angular project from 5 yo 6 i also face this ( location.back() ) problem.
I run all location.back inside zone.run but it does not solve this problem.
@wilker7ribeiro  is right about zone.
As they are expecting that all event are patched in zonejs
but i guess there was some issue in patching onpopstate event

Temp Solution:
I just edited /node_modules/@angular/router/esm5/router.js and revert [this change](https://github.com/angular/angular/compare/5.2.3...5.2.4#diff-a4a795ab9a55a8e79f03808ab30e6d39L72) 
and it works

**Comment by allensutton** [comment]
I think I have a related issue.  I created a PWA using Angular 6 and Material Design (not sure if any of that matters, I'm new at this), then realized that I needed native geolocation services to track the user's location in the background, when the screen is locked, etc. so I decided to try and migrate the app to a hybrid Angular/Cordova solution.  I get to the login screen, log in fine, get to the main page, which is like a dashboard with links on the sidenav panel.  If those links are clicked (which are child routes), the app can't find them.  I had originally imported BrowserAnimationsModule and didn't end up using them for anything.  I figured this might be the cause of the issue based on what I read on this thread, but removing the import statements for BrowserAnimationsModule breaks the whole thing, where not even the login screen comes up anymore -- even though I'm not using BAM anywhere?!

**Comment by rohitagni1993** [comment]
I think this is a better solution:

Write all the plugin calls and then when you want to trigger Angular to change the UI based on the plugin callback or something, write the code inside **NgZone**

Here's a snippet from my Angular Cordova app which struggled earlier to route to the logged in page once the plugin returned a success callback:

`this.cdvplugins.checkFingerprint().subscribe(
      (value) => {
        this.ngZone.run(() => {
          this.router.navigateByUrl('/myaccounts');
        });
      },
      (error) => {

      }`

**Comment by vano-maisuradze** [comment]
Thanks @rohitagni1993  NgZone fixed the issue.
`this.ngZone.run(() => {
 this.router.navigate(['your-route']);
});`

**Comment by dann95** [comment]
@Billskate it works so fine, damnnnnnnnn

**Comment by SergiyDagayev** [comment]
@Billskate, thank you for your solution! It fixed my Angular7.2+Cordova8.1 project.

**Comment by Jesusquinto** [comment]
> Thanks @rohitagni1993 NgZone fixed the issue.
> `this.ngZone.run(() => { this.router.navigate(['your-route']); });`
Thanks!!!, that work for me

**Comment by Jesusquinto** [comment]
@Billskate, thanks, that work for me!

**Comment by Jesusquinto** [comment]
@rohitagni1993 Thanks!, that work for me!

**Comment by lambirou** [comment]
@Billskate , thank you for your solution! It fixed for me Angular7.2 & Cordova8.1

**Comment by EenPOOL** [comment]
I don't really know what I'm doing wrong... 

The `ngZone.run(() => {})` runs fine, but it looks like overkill also it's annoying that one time you need to add it and another time it's not necessary... (also annoying because website & cordova app are sharing the same base code)

So I'm trying the @Billskate's aproach! But it doesn't look to have any effect  :confused: :tired_face:

Any suggestions? 
(ps: I have already tried to change script order...)

Angular 6.1 & Cordova 9 (iOS 4.5.5 & android 7.7.1)
``` html
<body>
  <app-root></app-root>
  <script>
    window.addEventListener = function () {
      (window.EventTarget || Window).prototype.addEventListener.apply(this, arguments);
    };
    window.removeEventListener = function () {
      (window.EventTarget || Window).prototype.removeEventListener.apply(this, arguments);
    };
    document.addEventListener = function () {
      (window.EventTarget || Document).prototype.addEventListener.apply(this, arguments);
    };
    document.removeEventListener = function () {
      (window.EventTarget || Document).prototype.removeEventListener.apply(this, arguments);
    };
  </script>
  <script src="cordova.js"></script>
  <script type="text/javascript" src="runtime.js"></script>
  <script type="text/javascript" src="polyfills.js"></script>
  <script type="text/javascript" src="styles.js"></script>
  <script type="text/javascript" src="scripts.js"></script>
  <script type="text/javascript" src="vendor.js"></script>
  <script type="text/javascript" src="main.js"></script>
</body>
```

**Comment by mjonas4356** [comment]
Hi, cordova.js must be after polyfills.js

**Comment by Billskate** [comment]
@EenPOOL Sorry, I don't have any ios4.5.5 or android 7.7.1 device, so I can't do any test for you.
You can try @mjonas4356 solution first.

**Comment by dann95** [comment]
@Billskate it works fine on IOS.

**Comment by EenPOOL** [comment]
Ok, still trying some stuff...

Meanwhile did anyone saw this possible solutions?
[https://github.com/angular/zone.js/blob/master/NON-STANDARD-APIS.md#others](https://github.com/angular/zone.js/blob/master/NON-STANDARD-APIS.md#others)

If I'm understand it well... zone.js created a patch for cordova and even other technologies! 😲 

```html
patch cordova.exec API

cordova.exec(success, error, service, action, args);

success and error will be patched with Zone.wrap.

to load the patch, you should load in the following order.

<script src="zone.js"></script>
<script src="cordova.js"></script>
<script src="zone-patch-cordova.js"></script>
```

**Comment by nj-coder** [comment]
Hi, is this issue solved or do we still have it? I cannot figure a way with any of the workarounds.

**Comment by Ganeshpoloju** [comment]
@Billskate works for me with angular8 also, thanks!

**Comment by nj-coder** [comment]
@Ganeshpoloju did it work in ios as well?

**Comment by nj-coder** [comment]
> Ok, still trying some stuff...
> 
> Meanwhile did anyone saw this possible solutions?
> https://github.com/angular/zone.js/blob/master/NON-STANDARD-APIS.md#others
> 
> If I'm understand it well... zone.js created a patch for cordova and even other technologies! 😲
> 
> ```
> patch cordova.exec API
> 
> cordova.exec(success, error, service, action, args);
> 
> success and error will be patched with Zone.wrap.
> 
> to load the patch, you should load in the following order.
> 
> <script src="zone.js"></script>
> <script src="cordova.js"></script>
> <script src="zone-patch-cordova.js"></script>
> ```

@EenPOOL did you get it working for ios and android?

**Comment by antman2** [comment]
Thank you so much, @Billskate, your Javascript patch work for us.

We had an Android/iOS Cordova app with a header/footer and main content inside a child route. It works just fine on desktop browser but wouldn't navigate back inside Cordova on Android and iOS.

Watching debug droppings you could see that when invoking Location.back() from Page2 it would progress through the Angular lifecycle as far as Page1.constructor() but didn't call Page1.ngOnInit(), so the content area remained blank. Clicking another button on the footer, such as to display a slide out, would resume the lifecycle at Page1.ngOnInit() and render the missing content.

**Comment by efficientmedia** [comment]
@Billskate works for me, thanks!

**Comment by ivancduran** [comment]
Hello everyone, this problem persist on angular 8 and 9, I'm using capacitor (with some cordova plugins, like the screen rotation) and I lost a lot of hours looking for a solutions and making code around zone.js, then I tried with the @Billskate patch and work very well for me.

So we need to look for a definitive fix, thank you everyone this old post solved my life today.

**Comment by yktan8** [comment]
I tried patch from @Billskate on Angular 9 and Cordova 9 and it works perfect!

**Comment by katzenjunge** [comment]
This is still an issue as of November 2020. The patch works, but it really should be properly fixed.

**Comment by abbasharoon** [comment]
As of now, Issue still exists in Angular 11 with Cordova 10. @Billskate solution worked but took a lot of time debugging the issue and finding the solution here. If it's not gonna be patched anytime soon then at least an error should be thrown in console.

**Comment by brian-weasner** [comment]
It looks like this is also happening for Capacitor applications that use some Cordova plugins.
Simply added this to the top of my `main.ts` and now my `location.back()` calls are working as expected within my capacitor app.

**Comment by alxhub** [comment]
I believe this is fixed by the `zone.js` patch for cordova. Is the patch not working for anyone?
